src/clients/c++/simple_shm_client.cc

```diff
@@ -0,0 +1,374 @@
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create status context: " << err << std::endl;
+    exit(1);
+  }
+
+  ni::ServerStatus server_status;
+  err = status_ctx->GetServerStatus(&server_status);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to get status: " << err << std::endl;
+    exit(1);
+  }
+
+  std::cout << "Status for model " << model_name << ":" << std::endl;
+  std::cout << server_status.DebugString() << std::endl;
+
+  // Create the inference context for the model.
+  std::unique_ptr<nic::InferContext> infer_ctx;
+  if (protocol == "http") {
+    err = nic::InferHttpContext::Create(
+        &infer_ctx, url, http_headers, model_name, -1 /* model_version */,
+        verbose);
+  } else if (protocol == "grpc") {
+    err = nic::InferGrpcContext::Create(
+        &infer_ctx, url, model_name, -1 /* model_version */, verbose);
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create inference context: " << err
+              << std::endl;
+    exit(1);
+  }
+
+  // Set the context options to do batch-size 1 requests. Also request
+  // that all output tensors be returned.
+  std::unique_ptr<nic::InferContext::Options> options;
+  FAIL_IF_ERR(
+      nic::InferContext::Options::Create(&options),
+      "unable to create inference options");
+
+  options->SetBatchSize(1);
+  for (const auto& output : infer_ctx->Outputs()) {
+    options->AddRawResult(output);
+  }
+
+  FAIL_IF_ERR(
+      infer_ctx->SetRunOptions(*options), "unable to set inference options");
+
+  // Create the data for the two input tensors. Initialize the first
+  // to unique integers and the second to all ones.
+  std::vector<int32_t> input0_data(16);
+  std::vector<int32_t> input1_data(16);
+  std::vector<int32_t> output0_data(16);
+  std::vector<int32_t> output1_data(16);
+  for (size_t i = 0; i < 16; ++i) {
+    input0_data[i] = i;
+    input1_data[i] = 1;
+  }
+
+  // Initialize the inputs with the data.
+  std::shared_ptr<nic::InferContext::Input> input0, input1;
+  std::shared_ptr<nic::InferContext::Output> output0, output1;
+  FAIL_IF_ERR(infer_ctx->GetInput("INPUT0", &input0), "unable to get INPUT0");
+  FAIL_IF_ERR(infer_ctx->GetInput("INPUT1", &input1), "unable to get INPUT1");
+  FAIL_IF_ERR(
+      infer_ctx->GetOutput("OUTPUT0", &output0), "unable to get OUTPUT0");
+  FAIL_IF_ERR(
+      infer_ctx->GetOutput("OUTPUT1", &output1), "unable to get OUTPUT1");
+
+  FAIL_IF_ERR(input0->Reset(), "unable to reset INPUT0");
+  FAIL_IF_ERR(input1->Reset(), "unable to reset INPUT1");
+
+  size_t input_byte_size = 16 * sizeof(int32_t);
+  size_t output_byte_size = 16 * sizeof(int32_t);
+  std::string shm_key;
+  int shm_fd, res;
+  void* shm_addr;
+
+  // Create  Input0 and Input1 in Shared Memory
+  // get shared memory file descriptor (NOT a file)
+  for (size_t n = 0; n < 2; n++) {
```

Comment:
Opening a shared memory region and memmapping a certain size should be encapsulated in a function.

Reply:
Will do